src/wallet/wallet.cpp

```diff
@@ -4567,3 +4567,287 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
 
     return ret;
 }
+
+bool CWallet::MigrateToDescriptor(bilingual_str& error, std::vector<bilingual_str>& warnings)
+{
+    LOCK(cs_wallet);
+
+    WalletLogPrintf("Migrating wallet from Legacy to Descriptor.\n");
+
+    LegacyScriptPubKeyMan* legacy_spkm = GetLegacyScriptPubKeyMan();
+    if (!legacy_spkm) {
+        error = _("Error: This wallet is already a descriptor wallet");
+        return false;
+    }
+
+    // Get all of the records for DB type migration
+    std::unique_ptr<DatabaseBatch> batch = m_database->MakeBatch();
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> records;
+    if (!batch->StartCursor()) {
+        error = _("Error: Unable to begin reading all records in the database");
+        return false;
+    }
+    bool complete = false;
+    while (true) {
+        CDataStream ss_key(SER_DISK, CLIENT_VERSION);
+        CDataStream ss_value(SER_DISK, CLIENT_VERSION);
+        bool ret = batch->ReadAtCursor(ss_key, ss_value, complete);
+        if (!ret) {
+            break;
+        }
+        std::vector<unsigned char> key(ss_key.begin(), ss_key.end());
+        std::vector<unsigned char> value(ss_value.begin(), ss_value.end());
+        records[key] = value;
+    }
+    batch->CloseCursor();
+    batch.reset();
+    if (!complete) {
+        error = _("Error: Unable to read all records in the database");
+        return false;
+    }
+
+    // Make a backup of the DB
+    std::string backup_path = strprintf("%s-%d.legacy.bak", m_database->Filename(), GetTime());
+    if (!BackupWallet(backup_path)) {
+        error = _("Error: Unable to make a backup of your wallet");
+        return false;
+    }
+
+    // Remove this wallet from the list of wallets while we mess with the database
+    // This is a workaround to prevent MaybeCompactWalletDB from accessing m_database while we are deleting and changing it
+    std::shared_ptr<CWallet> this_wallet{nullptr};
+    {
+        LOCK(cs_wallets);
+        std::vector<std::shared_ptr<CWallet>>::iterator i = std::find_if(vpwallets.begin(), vpwallets.end(), [this](std::shared_ptr<CWallet> w) {
+                return w.get() == this;
+            });
+        assert(i != vpwallets.end());
+        this_wallet = *i;
+        vpwallets.erase(i);
+    }
+
+    // Close this database and delete the file
+    fs::path db_path = fs::path(m_database->Filename());
+    fs::path db_dir = db_path.branch_path();
+    std::string db_filename = db_path.leaf().string();
+    m_database->Close();
+    fs::remove(db_path);
```

Comment:
This looks scary to me. Let's not delete the old wallet, but rename it. You know, just in case something with the migration didn't go well it's good to have a backup.

Reply:
~20 lines earlier we make a backup of the wallet. So there is still a backup and we're just removing the original after it is backed up.

Reply:
Okay, sounds good to me then.